Fix `BindingsArray` and `ObservablesArray` `reshape` to behave like NumPy

ihincks opened the pull request:
### Summary

Fixes #11606 by allowing both of the formats `reshape(1,2,3)` and `reshape((1,2,3))`.

### Details and comments

ihincks committed:
Fix BindingsArray.reshape and ObservablesArray.reshape to behave like NumPy

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -304,7 +304,7 @@ def reshape(self, shape: int | Iterable[int]) -> BindingsArray:
         """
-        shape = (shape, -1) if isinstance(shape, int) else (*shape, -1)
+        shape = shape_tuple(shape, -1)
         if np.prod(shape[:-1]).astype(int) != self.size:
```
```suggestion
        if np.prod(shape[:-1], dtype=int) != self.size:
```

chriseclectic reviewed (commented):
If they aren't already there can you add tests like:

```python
    def test_reshape(self):
        """Test reshape method"""
        bases = qi.pauli_basis(2)
        labels = np.array(bases.to_labels(), dtype=object)
        obs = ObservablesArray(qi.pauli_basis(2))

        def various_formats(shape):
            # call reshape with a single argument
            yield [shape]
            yield [(-1,) + shape[1:]]
            yield [np.array(shape)]
            yield [list(shape)]
            yield [list(map(np.int64, shape))]
            yield [tuple(map(np.int64, shape))]

            # call reshape with multiple arguments
            yield shape
            yield np.array(shape)
            yield list(shape)
            yield list(map(np.int64, shape))
            yield tuple(map(np.int64, shape))

        for shape in [(16,), (4, 4), (2, 4, 2), (2, 2, 2, 2), (1, 8, 1, 2)]:
            with self.subTest(shape):
                for input_shape in various_formats(shape):
                    obs_rs = obs.reshape(*input_shape)
                    self.assertEqual(obs_rs.shape, shape)
                    labels_rs = labels.reshape(shape)
                    for idx in np.ndindex(shape):
                        self.assertEqual(
                            obs_rs[idx],
                            {labels_rs[idx]: 1},
                            msg=f"failed for shape {shape} with input format {input_shape}",
                        )
```

```python
def test_reshape(self):
        """Test reshape"""
        with self.subTest("reshape"):
            vals = np.linspace(0, 1, 300).reshape((2, 3, 50))
            ba = BindingsArray(vals)
            reshape_ba = ba.reshape((3, 2))
            self.assertEqual(reshape_ba.num_parameters, 50)
            self.assertEqual(reshape_ba.ndim, 2)
            self.assertEqual(reshape_ba.shape, (3, 2))
            self.assertEqual(reshape_ba.size, 6)
            self.assertEqual(reshape_ba.kwvals, {})
            reshape_vals = vals.reshape((3, 2, 50))
            np.testing.assert_allclose(reshape_ba.vals, reshape_vals.reshape((1, 3, 2, 50)))

            bound_circuits = list(reshape_ba.bind_all(self.circuit).ravel())
            self.assertEqual(len(bound_circuits), 6)
            self.assertEqual(bound_circuits[0], self.circuit.assign_parameters(reshape_vals[0, 0]))
            self.assertEqual(bound_circuits[1], self.circuit.assign_parameters(reshape_vals[0, 1]))
            self.assertEqual(bound_circuits[2], self.circuit.assign_parameters(reshape_vals[1, 0]))
            self.assertEqual(bound_circuits[3], self.circuit.assign_parameters(reshape_vals[1, 1]))
            self.assertEqual(bound_circuits[4], self.circuit.assign_parameters(reshape_vals[2, 0]))
            self.assertEqual(bound_circuits[5], self.circuit.assign_parameters(reshape_vals[2, 1]))

        with self.subTest("flatten"):
            vals = np.linspace(0, 1, 300).reshape((2, 3, 50))
            ba = BindingsArray(vals)
            reshape_ba = ba.reshape(6)
            self.assertEqual(reshape_ba.num_parameters, 50)
            self.assertEqual(reshape_ba.ndim, 1)
            self.assertEqual(reshape_ba.shape, (6,))
            self.assertEqual(reshape_ba.size, 6)
            self.assertEqual(reshape_ba.kwvals, {})
            reshape_vals = vals.reshape(-1, 50)
            np.testing.assert_allclose(reshape_ba.vals, reshape_vals.reshape((1, 6, 50)))

            bound_circuits = list(reshape_ba.bind_all(self.circuit).ravel())
            self.assertEqual(len(bound_circuits), 6)
            for i in range(6):
                self.assertEqual(bound_circuits[i], self.circuit.assign_parameters(reshape_vals[i]))

        with self.subTest("various_formats"):
            ba = BindingsArray(
                [np.empty((16, 7)), np.empty((16, 3))],
                {Parameter("a"): np.empty(16), (Parameter("b"), Parameter("c")): np.empty((16, 2))},
            )

            def various_formats(shape):
                # call reshape with a single argument
                yield [shape]
                yield [(-1,) + shape[1:]]
                yield [np.array(shape)]
                yield [list(shape)]
                yield [list(map(np.int64, shape))]
                yield [tuple(map(np.int64, shape))]

                # call reshape with multiple arguments
                yield shape
                yield np.array(shape)
                yield list(shape)
                yield list(map(np.int64, shape))
                yield tuple(map(np.int64, shape))

            for shape in [(16,), (4, 4), (2, 4, 2), (2, 2, 2, 2), (1, 8, 1, 2)]:
                for input_shape in various_formats(shape):
                    reshaped_ba = ba.reshape(input_shape)
                    self.assertEqual(reshaped_ba.shape, shape)
```

ihincks committed:
improve reshaping tests to be more  comprehensive

ihincks commented:
Thanks, I've switched to the tests you pasted which are more comprehensive. They also caught a bug in `BindingsArray.reshape` where it wasn't dealing with `-1` values properly.

ihincks committed:
Fix bug with non-empty parameters when no kwals present

t-imamichi reviewed (dismissed):
LGTM

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -218,12 +218,16 @@ def as_array(self, parameters: Iterable[Parameter | str] | None = None) -> np.nd
                 ret[..., pos : pos + size] = arr
                 pos += size
-        elif self.kwvals:
+        else:
```
Ive been running into a lot of issues testing code with this method, and this change in particular seems to make existing workflows not possible.

The general issues is BindingsArray currently allows mixing and matching between named and unnamed parameter values, but then this array method currently wants only the named parameters to construct an array.

A typical workflow that might want this would be to use a circuits parameters to ensure ordering of this array. Eg
```
param_vals = bindings.as_array(circuit.parameters)
```
But currently this will error unless the bindings is all kwvals. 

I can think of two solutions for this issue:

1. We remove vals entirely from BindingsArray so all parameters have to be named. This would probably simplify the implementation, but it changes the types that can be converted to BindingsArrays by its `coerce` method.  In particular it removes the ability to initialize from an `np.array`. `EstimatorPub` could still coerce though since it also has access to the circuits so could use them in the coercion for a raw array input for param values.
2. We keep vals and kwvals in the init, but don't allow mixing the two. A BindingsArray has to be only defined with 1 or more vals arrays, or only defined with one or more kwvals arrarys, and if you try and initialize or coerce with a mix of both it errors. In this case methods like `as_array` in the case of only vals could only be able to validate that the number of parameters is correct, but ignore the names of the parameters, while the kwval case would validate as is done here.

ihincks replied:
I'd like to go forward with 1.

chriseclectic reviewed (changes requested):
I am still having issues in downstream code with the implementation of `as_array` in this PR. So I suggest either hodling off on merging until this can be resolved, or reverting these changes and opening a new PR to fix `BindingsArray.as_array` and merging this PR just as a fix to the `reshape` method (as stated in the original PR name).

ihincks committed:
Remove as_array() method

This method doesn't really serve an obvious purpose.
It would make more sence to have an nd-enumerate-like method to loop over parameter value sets instead.

ihincks commented:
I've removed `as_array()` altogether because it doesn't serve an obvious purpose: it's required only in very specific use cases.

chriseclectic reviewed (approved):
@ihincks I think `as_array` is still very useful, because without it if a user wants to combine all parameters from multiple kwvals into a single broadcasted array this is quite involved and easy to mess up. Could you add it back in in the PR you open to remove the `vals`?
